lib/IDE/ExprContextAnalysis.cpp

```diff
@@ -285,40 +285,59 @@ bool swift::ide::removeCodeCompletionExpr(ASTContext &Ctx, Expr *&expr) {
 
-Type swift::ide::getReturnTypeFromContext(const DeclContext *DC) {
+void swift::ide::collectPossibleReturnTypesFromContext(
+    DeclContext *DC, SmallVectorImpl<Type> &candidates) {
   if (auto FD = dyn_cast<AbstractFunctionDecl>(DC)) {
     auto Ty = FD->getInterfaceType();
     if (FD->getDeclContext()->isTypeContext())
       Ty = FD->getMethodInterfaceType();
-    if (auto FT = Ty->getAs<AnyFunctionType>())
-      return DC->mapTypeIntoContext(FT->getResult());
-  } else if (auto ACE = dyn_cast<AbstractClosureExpr>(DC)) {
-    if (ACE->getType() && !ACE->getType()->hasError())
-      return ACE->getResultType();
+    if (auto FT = Ty->getAs<AnyFunctionType>()) {
+      candidates.push_back(DC->mapTypeIntoContext(FT->getResult()));
+    }
+  }
+
+  if (auto ACE = dyn_cast<AbstractClosureExpr>(DC)) {
+    // Use the type checked type if it has.
+    if (ACE->getType() && !ACE->getType()->hasError() &&
+        !ACE->getResultType()->hasUnresolvedType()) {
+      candidates.push_back(ACE->getResultType());
+      return;
+    }
+
     if (auto CE = dyn_cast<ClosureExpr>(ACE)) {
       if (CE->hasExplicitResultType()) {
+        // If the closure has a explicit return type, use it.
         if (auto ty = CE->getExplicitResultType()) {
-          return ty;
+          candidates.push_back(ty);
+        } else {
+          auto typeLoc = TypeLoc{CE->getExplicitResultTypeRepr()};
+          if (!swift::performTypeLocChecking(
+                  DC->getASTContext(), typeLoc, /*isSILMode=*/false,
+                  /*isSILType=*/false, DC->getGenericEnvironmentOfContext(),
+                  const_cast<DeclContext *>(DC), /*diagnostics=*/false)) {
+            candidates.push_back(typeLoc.getType());
+          }
         }
-
-        auto typeLoc = TypeLoc{CE->getExplicitResultTypeRepr()};
-        if (swift::performTypeLocChecking(DC->getASTContext(),
-                                          typeLoc,
-                                          /*isSILMode*/ false,
-                                          /*isSILType*/ false,
-                                          DC->getGenericEnvironmentOfContext(),
-                                          const_cast<DeclContext *>(DC),
-                                          /*diagnostics*/ false)) {
-          return Type();
+      } else {
+        // Otherwise, check the context type of the closure.
+        ExprContextInfo closureCtxInfo(CE->getParent(), CE);
+        for (auto closureTy : closureCtxInfo.getPossibleTypes()) {
+          if (auto funcTy = closureTy->getAs<AnyFunctionType>())
+            candidates.push_back(funcTy->getResult());
         }
-        return typeLoc.getType();
       }
     }
+
+    // Even if the type checked type has unresolved types, it's better than
+    // nothing.
+    if (candidates.empty() && ACE->getType() && !ACE->getType()->hasError()) {
+      candidates.push_back(ACE->getResultType());
+      return;
```

Comment:
Unnecessasry `return`.  Alternatively, maybe we should early-return everywhere above that adds candidates? It's not obvious that all the above cases are mutually exclusive.